### PR TITLE
Only fail relocation target shard if failing source shard is a primary

### LINKED ISSUE #16144: Failing replica relocation source should not fail replica relocation target
If the relocation source fails during the relocation of a shard from one node to another, the relocation target is currently failed as well. For replica shards this is not necessary, however, as the actual shard recovery of the relocation target is done via the primary shard.

### PR DESCRIPTION
**Pull request by ywelsch**
If the relocation source fails during the relocation of a shard from one node to another, the
relocation target is currently failed as well. For replica shards this is not necessary,
however, as the actual shard recovery of the relocation target is done via the primary shard.

Closes #16144

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/CancelAllocationCommand.java [hunk 1]
```diff
@@ -156,17 +156,30 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
...
                         while (initializingNode.hasNext()) {
                             ShardRouting initializingShardRouting = initializingNode.next();
                             if (initializingShardRouting.isRelocationTargetOf(shardRouting)) {
-                                initializingNode.remove();
+                                if (shardRouting.primary()) {
+                                    initializingNode.remove();
+                                } else {
+                                    // relocation source is non-primary, mark this shard as initializing without relocation source
+                                    removedRelocationSourceOnRelocationTarget = true;
+                                    initializingNode.removeRelocationSource();
```
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/CancelAllocationCommand.java]: Why not call `initializingNode.removeRelocationSource()` in the `if (removedRelocationSourceOnRelocationTarget)` conditional below?

**Comment by ywelsch** [comment]
@s1monw can you have a look at this one?

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java [hunk 2]
```diff
@@ -598,14 +599,24 @@ private boolean applyFailedShard(RoutingAllocation allocation, ShardRouting fail
...
+                                initializingNode.removeRelocationSource();
+                            }
+                            break;
                         }
                     }
                 }
             }
-
-            matchedNode.moveToUnassigned(unassignedInfo);
+            if (removedRelocationSourceOnRelocationTarget) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: hmm I am not sure I understand why this is skipped. The `matchedNode` is the Routing node that holds the shard that is failed. So we have to move it to unassigned anyway no? I must be missing something - this is a confusion method :)

**Comment by s1monw** [comment]
I left one comment @ywelsch - looks good otherwise

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java (hunk 2 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: The current logic in this method is a bit weird.
When a shard is found that matched `failedShard` we first call `matchedNode.remove()` (earlier up in the method, not shown by Github diff here). This means that the shard gets removed. This makes sense for example if the failed shard is a relocation target. If it is a started shard, however, we need to readd it as unassigned shard, which happens two lines below this comment.

The case we want to cover in this PR is that we remove relocation source (not add it back as unassigned shard) but keep the relocation target.

Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: I was a bit hesitant on refactoring the existing flow of the method. I can give that a try though if you think it helps.

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: alright now I got it! Can we rename `removedRelocationSourceOnRelocationTarget` to `addAsUnassigned` initalized to `true` and set it to false if we removed the source only. just find something that reflects the situation and maybe add a comment that explains this?

Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: I've pushed f7b685c following your suggestion.

**Comment by s1monw** [comment]
much easier to understand - LGTM thanks @ywelsch

**Comment by ywelsch** [comment]
Thanks for reviewing @s1monw!

**Pull request merged**